src/outside/grammar_transformation.cc

```diff
@@ -0,0 +1,967 @@
+    }
+  }
+
+  if (warn_missing_nts->size() > 0) {
+    std::string msg = std::string(
+      "You requested outside grammar generation and\nreporting results for "
+      "the following non-terminals, which do NOT exist in your grammar '" +
+      *this->name + "':\n");
+    for (std::list<std::string>::iterator i = warn_missing_nts->begin();
+         i != warn_missing_nts->end(); ++i) {
+      msg += "  '" + *i + "'";
+      if (next(i) != warn_missing_nts->end()) {
+        msg += "\n";
+      }
+    }
+    throw LogError(this->location, msg);
+  }
+}
+
+// traverses the grammar and collect all algebra function names used
+// such that after traversal, we can ask if an algebra function,
+// given its name, is actually part of the grammar
+struct AlgfctUsedInGrammar : public Visitor {
+ private:
+  std::set<std::string> used_algfct;
+  bool is_traversed = false;
+
+ public:
+  void visit(Alt::Base &a) {
+    Alt::Simple *as = dynamic_cast<Alt::Simple*>(&a);
+    if (as) {
+      used_algfct.insert(*as->name);
+    }
+    is_traversed = true;
+  }
+
+  bool is_used(std::string *algfct_name) {
+    assert(is_traversed);
+    return used_algfct.find(*algfct_name) != used_algfct.end();
+  }
+};
+
+bool is_terminal_type(Type::Base *t) {
+  if ((t->is(Type::ALPHABET)) || (t->is(Type::VOID)) ||
+      (t->is(Type::REALVOID)) || (t->is(Type::CHAR)) ||
+      (t->is(Type::STRING)) || (t->is(Type::BOOL)) || (t->is(Type::SEQ)) ||
+      (t->is(Type::SUBSEQ)) || (t->is(Type::INT)) || (t->is(Type::FLOAT)) ||
+      (t->is(Type::RATIONAL))) {
+    return true;
+  }
+  if ((t->is(Type::BIGINT)) || (t->is(Type::SHAPE)) || (t->is(Type::INTEGER))) {
+    return false;
+  }
+  if ((t->is(Type::EXTERNAL)) || (t->is(Type::TUPLEDEF))) {
+    return false;
+  }
+  if (t->is(Type::USAGE)) {
+    return (t->is_terminal()) && (is_terminal_type(t->simple()));
+  }
+  if (t->is(Type::MULTI)) {
+    Type::Multi *tm = dynamic_cast<Type::Multi*>(t);
+    for (std::list<Type::Base*>::const_iterator i = tm->types().begin();
+         i != tm->types().end(); ++i) {
+      if (!is_terminal_type(*i)) {
+        return false;
+      }
+    }
+    return true;
+  }
+  if (t->is(Type::SINGLE)) {
+    return t->is_terminal();
+  }
+  if (t->is(Type::REFERENCABLE)) {  // a pointer to an object
+    return (dynamic_cast<Type::Referencable*>(t)->base->is_terminal());
+  }
+  if (t->is(Type::SIGNATURE) ||  // a basic ADP component
+      t->is(Type::TABLE) ||  // a DP matrix to store non-terminal results
+      t->is(Type::LIST) ||  // the answer list of an algebra function
+      t->is(Type::TUPLE) ||  // a pair of results
+      t->is(Type::CHOICE) ||  // flag an algebra function as a choice function
+      t->is(Type::RANGE) ||  // codegen, set of begin/end iterators
+      t->is(Type::SIZE)  // internal type for code generation = unsigned int
+      ) {
+    return false;
+  }
+  if ((dynamic_cast<Type::Name*>(t)) ||  // the name of a new user defined tuple
```

Comment:
I don't really understand the purpose of this `if` statement. Dynamic-casting the `Type::Base* t` to pointers to derived classes of this base type and checking them like this in an `if` statement will always evaluate to `true` if `t != nullptr`, regardless of the type `t` is casted into. If  `t == nullptr`, all casts inside the `if` statement would evaluate to `false`. If a certain cast wouldn't be allowed/possible, `dynamic_cast` would cause a compiler error. So essentially this `if` statement is just checking if  `t != nullptr` (or just `if (t)` simplified) , which I think should either be the first check inside `is_terminal_type` so we don't try to dereference a null pointer here, or just left out completely if we can be sure that `t` is never going to be a null pointer.

Reply:
My mistake. `dynamic_cast` actually does return `nullptr` if a cast from a Base class to derived class fails due to the derived class being the wrong type. I wrongly assumed that this would also cause an error. So this `if` statement now does make sense and can be used to figure out if `t` is of a specific derived class.

However, you're also rightfully pointed out that `Type::Name` isn't a derived type from `Type::Base`. So for that type this cast probably doesn't work as intended as causes undefined behaviour according to [cppreference](https://en.cppreference.com/w/cpp/language/dynamic_cast). So here we would either need to make `Type::Name` a derived type of `Type::Base` or remove the check for `Type::Name` in the `if` statement.